src/WebSerialLite.cpp

```diff
@@ -0,0 +1,183 @@
+#include "WebSerialLite.h"
+
+#include "wslp.h"
+
+#define WSL_MAGIC_BYTE_1     0xAB
+#define WSL_MAGIC_BYTE_2     0xCD
+#define WSL_MAX_MESSAGE_SIZE UINT16_MAX
+
+typedef enum {
+  WSL_WRITE_ROW = 0x01,
+  WSL_MESSAGE = 0x02,
+  WSL_PING = 0x03,
+  WSL_PONG = 0x04,
+} WSLPacketType;
+
+static const uint8_t PONG[] = {WSL_MAGIC_BYTE_1, WSL_MAGIC_BYTE_2, WSLPacketType::WSL_PONG};
+static const uint8_t HEAD[] = {
+  WSL_MAGIC_BYTE_1,             // Magic Bytes
+  WSL_MAGIC_BYTE_2,             // Magic Bytes
+  WSLPacketType::WSL_WRITE_ROW, // Packet Type (1 byte)
+  0x00,                         //
+  0x00,                         //
+  0x00,                         //
+  0x00,                         //
+  0x00,                         //
+  0x00,                         //
+  0x00,                         //
+  0x00,                         // Reserved (8 bytes)
+  0x00                          // Reserved (1 byte)
+};
+static const size_t WSL_PONG_LEN = sizeof(PONG) / sizeof(PONG[0]);
+static const size_t WSL_HEAD_LEN = sizeof(HEAD) / sizeof(HEAD[0]);
+static const size_t WSL_SIZE_LEN = sizeof(uint16_t);
+
+void WebSerialLiteClass::setAuthentication(const String& username, const String& password) {
+  _username = username;
+  _password = password;
+  _authenticate = !_username.isEmpty() && !_password.isEmpty();
+  if (_ws) {
+    _ws->setAuthentication(_username.c_str(), _password.c_str());
+  }
+}
+
+void WebSerialLiteClass::begin(AsyncWebServer* server, const char* url) {
+  _server = server;
+  _ws = new AsyncWebSocket("/wserial");
+
+  if (_authenticate) {
+    _ws->setAuthentication(_username, _password);
+  }
+
+  _server->on(url, HTTP_GET, [&](AsyncWebServerRequest* request) {
+    if (_authenticate) {
+      if (!request->authenticate(_username.c_str(), _password.c_str()))
+        return request->requestAuthentication();
+    }
+    AsyncWebServerResponse* response = request->beginResponse_P(200, "text/html", WEBSERIAL_HTML, sizeof(WEBSERIAL_HTML));
+    response->addHeader("Content-Encoding", "gzip");
+    request->send(response);
+  });
+
+  _ws->onEvent([&](AsyncWebSocket* server, AsyncWebSocketClient* client, AwsEventType type, void* arg, uint8_t* data, size_t len) -> void {
+    if (type == WS_EVT_CONNECT) {
+      client->setCloseClientOnQueueFull(false);
```

Comment:
FYI: explanations here: https://github.com/mathieucarbou/ESPAsyncWebServer/commit/da9cfe7f01e5a8cf4dd7d3ff484c6588db71a330

There is a tradeoff to make in async libs regarding the message queue and the behaviour to select when it overflows.: either disconnect the client and loose further messages, or discard the messages only.

The original lib from me-no-dev, and the ESPHome fork are discarding only.  So the client had to be cleaned up and the UI, if any, could not be hinted that there is an issue. But it is way faster and goes through less code cleanup and mutex.

The fork from yubox-node, which brought a lot of concurrency and stability fixes, was closing client.  So this allows the UI to react and recreate the connection, eventually layout. The issue is that with a high flow of messages (more than 20 per second) at one point the ESP crashes when trying to aquire a lock, and it can also create a loop, after the UI has re-created the connection, the issue can happen again.

WebSerial not having a lot of clients, and mostly used to stream logs, it is better to just discard messages to get more perf out of it.